tools/lint-buffer-holder/main.go

```diff
@@ -0,0 +1,354 @@
+		Fset:  l.fset,
+		Mode:  packages.LoadAllSyntax,
+		Tests: false,
+	}, pkgPaths...))
+	l.fillPackages(packages)
+
+	l.seen = map[types.Type]bool{}
+	for _, p := range packages {
+		for _, service := range services {
+			if p.Dir != must(filepath.Abs(filepath.Dir(service.protoPath))) {
+				continue
+			}
+			l.analyzeGRPCServer(p, service.name)
+		}
+	}
+}
+
+func (l *linter) fillPackages(pkgs []*packages.Package) {
+	l.packages = make(map[string]*packages.Package)
+
+	var visit func(*packages.Package)
+	visit = func(pkg *packages.Package) {
+		if _, seen := l.packages[pkg.PkgPath]; seen {
+			return
+		}
+		l.packages[pkg.PkgPath] = pkg
+
+		for _, imp := range pkg.Imports {
+			visit(imp)
+		}
+	}
+
+	for _, pkg := range pkgs {
+		visit(pkg)
+	}
+
+	l.contextContext = l.packages["context"].Types.Scope().Lookup("Context").(*types.TypeName).Type()
+	l.errorType = types.Universe.Lookup("error").(*types.TypeName).Type()
+	l.byteType = types.Universe.Lookup("byte").(*types.TypeName).Type()
+	l.messageWithBufferRef = l.packages["github.com/grafana/mimir/pkg/mimirpb"].Types.Scope().Lookup("MessageWithBufferRef").(*types.TypeName).Type().Underlying().(*types.Interface)
+	l.unsafeMutableString = l.packages["github.com/grafana/mimir/pkg/mimirpb"].Types.Scope().Lookup("UnsafeMutableString").(*types.TypeName).Type().(*types.Alias)
+}
+
+type grpcService struct {
+	protoPath string
+	name      string
+}
+
+// findGRPCServices finds all gRPC services in .proto files under ./pkg.
+// It returns an iterator that yields ServiceInfo for each service found.
+func findGRPCServices(dirs ...string) iter.Seq[grpcService] {
+	if len(dirs) == 0 {
+		dirs = []string{"./pkg"}
+	}
+	return func(yield func(grpcService) bool) {
+		// Walk the directory tree looking for .proto files
+		for _, dir := range dirs {
+			try(filepath.WalkDir(dir, func(path string, d os.DirEntry, err error) error {
+				if err != nil {
+					return err
+				}
+
+				// Skip directories and non-proto files
+				if d.IsDir() || !strings.HasSuffix(d.Name(), ".proto") {
+					return nil
+				}
+
+				// Parse the proto file for service definitions
+				services, err := parseProtoServices(path)
+				if err != nil {
+					return fmt.Errorf("failed to parse %s: %w", path, err)
+				}
+
+				// Yield each service found
+				for _, serviceName := range services {
+					if !yield(grpcService{
+						protoPath: path,
+						name:      serviceName,
+					}) {
+						return filepath.SkipAll
+					}
+				}
+
+				return nil
+			}))
+		}
+	}
+}
+
+// parseProtoServices extracts service names from a proto file
+func parseProtoServices(filePath string) ([]string, error) {
+	file, err := os.Open(filePath)
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	var services []string
+	scanner := bufio.NewScanner(file)
+
+	// Regular expression to match service declarations
+	// Matches: service ServiceName {
+	serviceRegex := regexp.MustCompile(`^\s*service\s+(\w+)\s*\{`)
```

Comment:
Rather than parsing the `.proto` files ourselves using regexes, is there some way we can use the existing Protobuf parser, Go reflection or gRPC metadata to do this?

For example, each gRPC service results in a generated `grpc.ServiceDesc` field (eg. [this one](https://github.com/grafana/mimir/blob/4b93609707986ed1fe8ff22ece048e1a0a9a55d8/pkg/storegateway/storegatewaypb/gateway.pb.go#L226) for the store-gateway) - could we look for those?